core/api/src/domain/ledger/ln-payment-state.ts

```diff
@@ -0,0 +1,110 @@
+  arr.reduce((count, currentValue) => {
+    return currentValue === valueToCount ? count + 1 : count
+  }, 0)
+
+const sum = <T>({ arr, propertyName }: { arr: T[]; propertyName: keyof T }): number =>
+  arr.reduce((sum, currentItem) => {
+    return sum + Number(currentItem[propertyName])
+  }, 0)
+
+const checkTxns = (
+  txns: LedgerTransaction<WalletCurrency>[],
+): true | InvalidLnPaymentTxnsBundleError => {
+  const validTypes = [
+    LedgerTransactionType.Payment,
+    LedgerTransactionType.LnFeeReimbursement,
+  ]
+  for (const txn of txns) {
+    if (!validTypes.includes(txn.type as (typeof validTypes)[number])) {
+      return new InvalidLnPaymentTxnsBundleError(`Invalid '${txn.type}' type found`)
+    }
+  }
+
+  if (txns.length === 0) return new InvalidLnPaymentTxnsBundleError("No txns in bundle")
+
+  return true
+}
+
+export const LnPaymentStateDeterminator = (
+  txns: LedgerTransaction<WalletCurrency>[],
+): LnPaymentStateDeterminator => {
+  const determine = (): LnPaymentState | InvalidLnPaymentTxnsBundleError => {
+    const check = checkTxns(txns)
+    if (check instanceof Error) return check
+
+    const txPendingStates = txns.map((txn) => txn.pendingConfirmation)
+    if (txPendingStates.includes(true)) {
+      const pendingCount = count({ arr: txPendingStates, valueToCount: true })
+
+      switch (true) {
+        case txns.length === 1:
+          return LnPaymentState.Pending
+
+        case txns.length % 2 === 1 && pendingCount === 1:
+          return LnPaymentState.PendingAfterRetry
+
+        default:
+          return new InvalidLnPaymentTxnsBundleError()
+      }
+    }
+
+    const txTypes = txns.map((txn) => txn.type)
+    const txPaymentCount = count({
+      arr: txTypes,
+      valueToCount: LedgerTransactionType.Payment,
+    })
+    if (txTypes.includes(LedgerTransactionType.LnFeeReimbursement)) {
+      const txReimburseCount = count({
+        arr: txTypes,
+        valueToCount: LedgerTransactionType.LnFeeReimbursement,
+      })
+      switch (true) {
+        case txPaymentCount === 1 && txReimburseCount === 1:
+          return LnPaymentState.SuccessWithReimbursement
+        case txPaymentCount > 1 && txReimburseCount === 1:
+          return LnPaymentState.SuccessWithReimbursementAfterRetry
+        default:
+          return new InvalidLnPaymentTxnsBundleError()
+      }
+    }
+
+    const sumTxAmounts =
+      sum({ arr: txns, propertyName: "debit" }) -
+      sum({ arr: txns, propertyName: "credit" })
+    const failedUsdPayment = !!txns.find((tx) => tx.lnMemo === FAILED_USD_MEMO)
+    switch (true) {
+      case txns.length === 1:
+        return LnPaymentState.Success
+
+      case txns.length === 2 && (sumTxAmounts === 0 || failedUsdPayment):
+        return LnPaymentState.Failed
+
+      case txns.length % 2 === 1:
+        return LnPaymentState.SuccessAfterRetry
+
+      case txns.length % 2 === 0 && sumTxAmounts === 0:
+        return LnPaymentState.FailedAfterRetry
```

Comment:
How would you formulate that conditional and why is it better than this? (maybe readability or robustness?)

Reply:
Ok I tried switching to using voids, but the issue is when we record a `FailedUsdPayment` back to the BTC wallet we don't void the original btc transaction since we don't go through medici for this operation. Not sure it's worth it here to change that behaviour and do a migration (again because of sensitivity of touching the ledger, and upcoming planned changes).